kimchi/src/snarky/checked_runner.rs

```diff
@@ -0,0 +1,424 @@
+//! The circuit-generation and witness-generation logic.
+
+use ark_ff::PrimeField;
+
+use crate::{
+    circuits::gate::CircuitGate,
+    curve::KimchiCurve,
+    loc,
+    snarky::{
+        boolean::Boolean,
+        constraint_system::{BasicSnarkyConstraint, KimchiConstraint, SnarkyConstraintSystem},
+        cvar::CVar,
+        traits::SnarkyType,
+    },
+};
+
+use super::constants::Constants;
+
+/// A wrapper around [BasicSnarkyConstraint] and [KimchiConstraintSystem] that allows for an optional label (for debugging).
+pub struct AnnotatedConstraint<F: PrimeField> {
+    annotation: Option<&'static str>,
+    constraint: Constraint<F>,
+}
+
+impl<F> AnnotatedConstraint<F>
+where
+    F: PrimeField,
+{
+    /// In witness generation, this checks if the constraint is satisfied by some witness values.
+    pub fn check_constraint(&self, env: &impl WitnessGeneration<F>) {
+        match &self.constraint {
+            Constraint::BasicSnarkyConstraint(c) => c.check_constraint(env),
+            Constraint::KimchiConstraint(c) => c.check_constraint(env),
+        }
+    }
+}
+
+/// An enum that wraps either a [BasicSnarkyConstraint] or a [KimchiConstraintSystem].
+// TODO: we should get rid of this once basic constraint system is gone
+pub enum Constraint<F: PrimeField> {
+    /// Old R1CS-like constraints.
+    BasicSnarkyConstraint(BasicSnarkyConstraint<CVar<F>>),
+
+    /// Custom gates in kimchi.
+    KimchiConstraint(KimchiConstraint<CVar<F>, F>),
+}
+
+/// The two ways to create a [SnarkyType].
+#[derive(Default)]
+pub enum TypeCreation {
+    /// This will create constraints in the constraint system to ensure that the type is valid.
+    /// This is done by calling [SnarkyType::check] on a snarky type.
+    #[default]
+    Checked,
+
+    /// This will create a new type without any constraints.
+    /// This should only be used internally when we know the type is already constrained to be valid.
+    Unsafe,
+}
+
+/// The mode in which the [RunState] is running.
```

Comment:
IIRC, the design of the OCaml snarky allows you to do both at once -- that is, to generate a witness while you're also building the constraint system. What do you think about enabling this here too? (Doesn't need to be now / in this PR.)

Reply:
tbh I don't like all the hidden modes and ways to run snarky, I think the API should remain pretty basic: compile, and then construct witness